Add documention on using alternative JVM vendor buildpacks. (#131)
paketo-buildpacks/paketo-website: content/docs/buildpacks/language-family-buildpacks/java.md

The resulting application image will be identical to that built in the Building with Maven example.

## About the JVM

The Java Buildpack uses the [BellSoft Liberica][liberica] implementations of the JRE and JDK. JVM installation is handled by the [BellSoft Liberica Buildpack][bp/bellsoft-liberica]. The JDK will be installed in the build container but only the JRE will be contributed to the application image.

> See the [homepage][bp/bellsoft-liberica] for the Bellsoft Liberica Buildpackfor a full set of configuration options.

Reviewer: ```suggestion
See the [homepage][bp/bellsoft-liberica] for the Bellsoft Liberica Buildpack for a full set of configuration options.
```
this was my typo from before but we might as well fix it while we are looking at it :)

### Runtime JVM Configuration

The Java Buildpack configures the JVM by setting `JAVA_TOOL_OPTIONS` in the JVM environment.

#### Memory Calculator

The Java Buildpack installs a component called the Memory Calculator which will configure JVM memory based on the resources available to the container at runtime. The calculated flags will be appended to `JAVA_TOOL_OPTIONS`.

#### Configuring JVM at Runtime

The runtime JVM can be configured in two ways:

1. Buildpack-provided runtime components including the Memory Calculator accept semantically named environment variables which are then used to derive `JAVA_TOOL_OPTIONS` flags. Examples include:
    * `BPL_JVM_HEAD_ROOM`
    * `BPL_JVM_LOADED_CLASS_COUNT`
    * `BPL_JVM_THREAD_COUNT`
2. Flags can be set directly at runtime with the `JAVA_TOOL_OPTIONS` environment variable. User-provided flags will be appended to buildpack-provided flags. If the user and a buildpack set the same flag, user-provided flags take precedence.

See the [homepage][bp/bellsoft-liberica] for the Bellsoft Liberica Buildpack for a full set of configuration options.

> ### Using Alternative JVM Vendors

Reviewer: ```suggestion
### Using Alternative JVMs
```

> By default, the [Paketo Java buildpack][bp/java] will use the Liberica JVM. The following alternative JVM vendors may be substituted in place of Liberica's JVM.

Reviewer: ```suggestion
By default, the [Paketo Java buildpack][bp/java] will use the Liberica JVM. The following Paketo JVM buildpacks may be used to substitute alternate JVM implemenations in place of Liberica's JVM.
```

Reviewer: nit: I got a bit thrown off by the way we are using the word vendor. I know it was used in the title of the associated issue but if we are going to standardize on terminology I would rather call this category of buildpacks "JVM buildpacks" instead of "JVM vendor buildpacks".

Author: I have seen the term vendor used for different variants of OpenJDK, but I'm fine with keeping it short. I'm sometimes too wordy.

* [Paketo Amazon Corretto Buildpack][bp/amazon-corretto]
* [Paketo Azul Zulu Buildpack][bp/azul-zulu]
* [Paketo BellSoft Liberica Buildpack - Default][bp/bellsoft-liberica]
* [Paketo Eclipse OpenJ9 Buildpack][bp/eclipse-openj9]
* [Paketo GraalVM Buildpack][bp/graalvm]

> * [Paketo SapMachine Buildpack][bp/sap-machine]

Reviewer: Maybe we should make this a two column table that associates JVM implementations with their respective buildpacks? I suggest this mostly because I think it would be nice to include links to the homepages of the JVMs themselves so folks can click through and read about the options.

Reviewer: looks great!

> To use an alternative JVM vendor, you will need to set two `--buildpack` arguments to `pack build`, one for the alternative JVM vendor you'd like to use and one for the Paketo Java buildpack (in that order). This works because while you end up with two JVM vendor buildpacks, the first one, the one you're specifying will claim the buildplan entries so the second one will end up being a noop and doing nothing.

Reviewer: ```suggestion
To use an alternative JVM, you will need to set two `--buildpack` arguments to `pack build`, one for the alternative JVM buildpack you'd like to use and one for the Paketo Java buildpack (in that order). This works because while you end up with two JVM buildpacks, the first one, the one you're specifying will claim the build plan entries so the second one will end up being a noop and doing nothing.
```

This example will switch in the Azul Zulu buildpack:

{{< code/copyable >}}
pack build samples/jar --buildpack gcr.io/paketo-buidpacks/azul-zulu --buildpack paketo-buildpacks/java`
{{< /code/copyable >}}

There is one drawback to this approach. When using the method above to specify an alternative JVM vendor buildpack, this alternate buildpack ends up running before the CA certs buildpack and therefore traffic from the alternate JVM vendor buildpack won’t trust any additional CA certs. This is not expected to impact many users because JVM buildpacks should reach out to URLs that have a cert signed by a known authority with a CA in the default system truststore.

> If you have customized your JVM vendor buildpack to download the JVM from a URL that uses a certificate not signed by a well-known CA, you can workaround this by specifying the CA certs buildpack to run first. This works because while you will end up with the CA certificates buildpack specified twice, the lifecycle is smart enough to drop the second one.

Reviewer: ```suggestion
If you have customized your JVM buildpack to download the JVM from a URL that uses a certificate not signed by a well-known CA, you can workaround this by specifying the CA certs buildpack to run first. This works because while you will end up with the CA certificates buildpack specified twice, the lifecycle is smart enough to drop the second one.
```
